### PR TITLE
[SQL] Optimization to pull aggregate filters into temporal filters where possible

### LINKED ISSUE #6655: [SQL] Optimize the use of temporal filter-like expressions in aggregates
```sql
SELECT SUM(CASE WHEN tt >= (NOW() - INTERVAL '24' HOUR) THEN 1 END) AS tc_24h,
            SUM(CASE WHEN tt >= (NOW() - INTERVAL '1' HOUR) THEN 1 END) AS tc_1h
FROM T
GROUP BY ...
```

Today this compiles into a cross-join between T and NOW which re-evaluates fully on every clock tick.
This can be compiled using a form of temporal filters. The re-evaluation will still require a (multi-)join on the group key, but at least the inputs of the join should only contain the changed rows of T.

### PR DESCRIPTION
**Pull request by mihaibudiu**
Fixes #6655

### Describe Manual Test Plan

I have manually checked the generated plans.

## Checklist

- [x] Unit tests added/updated
- [x] Documentation updated

Example program:
```sql
CREATE VIEW V AS SELECT
   id,
  SUM(CASE WHEN tt >= (NOW() - INTERVAL '24' HOUR) THEN 1 END) AS tc_24h,
  SUM(CASE WHEN tt >= (NOW() - INTERVAL '1' HOUR) THEN 1 END) AS tc_1h,
  MAX(tt) AS max_tt
FROM T
  WHERE tt BETWEEN (NOW() - INTERVAL '25' HOUR) AND NOW() AND id IS NOT NULL
  GROUP BY id;
```

This is implemented in a Calcite-level optimization which pulls the conditions gating the aggregates into filters, which are subsequently implemented as temporal filters where possible.

Plan before:

before

Plan after:

final

The plan has more joins, but the inputs to the joins come from Window operators, which produce much smaller deltas (only rows that changed since last clock tick).

### CODE REVIEW

**Comment by ryzhyk** [comment]
@mihaibudiu , what is the most general pattern this recognizes?

**Comment by mihaibudiu** [comment]
It is actually a composition of several independent rules, some of which used to exist.

- There is a pre-existing Calcite optimization which rewrites `SELECT AGG(CASE exp THEN 1)` into `SELECT AGG(1) FILTER (exp)` (this is a standard SQL extension that we support)
- What I have implemented here is a new optimization to convert such an aggregate whose filter contains an expression that depends on NOW; such an aggregate is rewritten as `SELECT AGG(1) WHERE (exp)`.  The harder part is that multiple aggregates that may have required a single pass need a JOIN on the group key to combine their results.
- The pre-existing filter code recognizes temporal filters in `WHERE`, no changes required there

**Comment by mihaibudiu** [comment]
I think all 3 rules are as general as possible, but if there will be more patterns that can be optimized, we'll reconsider.

**Comment by ryzhyk** [comment]
> I think all 3 rules are as general as possible, but if there will be more patterns that can be optimized, we'll reconsider.

That's nice

File: sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/calciteCompiler/optimizer/AggregateNowFilterRule.java [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+        final RelNode anchor = builder.build();
+
+        builder.push(anchor).push(filtered);
+        if (groupCount == 0) {
+            // Both sides produce exactly one row
+            builder.join(JoinRelType.INNER, builder.literal(true));
+        } else {
+            List<RexNode> conditions = new ArrayList<>();
+            for (int i = 0; i < groupCount; i++) {
+                // Use IS_NOT_DISTINCT_FROM for the JOIN condtioin to treat NULL keys as equal
```
Comment (mythical-fred) [inline sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/calciteCompiler/optimizer/AggregateNowFilterRule.java]: Typo: `condtioin` -> `condition`.

File: sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/calciteCompiler/optimizer/AggregateNowFilterRule.java [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+        final Aggregate aggregate = call.rel(0);
+        final Project project = call.rel(1);
+        if (aggregate.getGroupType() != Aggregate.Group.SIMPLE)
+            return;
+
+        final List<RexNode> projects = project.getProjects();
+        // The column containing the filter with NOW()
+        int filterColumn = -1;
+        for (AggregateCall aggCall : aggregate.getAggCallList()) {
+            if (aggCall.filterArg >= 0 && containsNow(projects.get(aggCall.filterArg))) {
```
Comment (mythical-fred) [inline sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/calciteCompiler/optimizer/AggregateNowFilterRule.java:119]: Question about non-temporal degradation: the trigger is "filter expression contains a `NOW()` call", but the downstream temporal-filter transformation only recognizes a restricted shape (e.g. `col >= NOW() - INTERVAL ...` or a conjunction thereof). If a user writes something like `SUM(CASE WHEN NOW() BETWEEN a AND b THEN 1 END) ... GROUP BY k`, this rule will still fire and produce a `LEFT JOIN` plus a `Filter` that still cross-joins with `NOW` -- likely strictly worse than the original single aggregate. Have you considered gating the rewrite on the condition actually being of the form a temporal-filter can consume, or is that expected to be rare enough not to matter in practice?

File: sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/streaming/StreamingTests.java [hunk 3]
```diff
@@ -2512,6 +2512,234 @@ create table T (
...
+                 2  |NULL    |NULL   | 2019-12-31 11:30:00 | -1""");
+        // A day later id 1's rows leave the WHERE band as well
+        ccs.step("INSERT INTO NOW VALUES('2020-01-02 13:00:00')", """
+                 id | tc_24h | tc_1h | max_tt              | weight
+                -----------------------------------------------------
+                 1  | 2      |NULL   | 2020-01-01 11:30:00 | -1""");
+    }
+
+    @Test
+    public void issue6655d() {
```
Comment (mythical-fred) [inline sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/streaming/StreamingTests.java:2722]: Small nit: this test verifies the outputs, but not that the rule was correctly *blocked*. Consider mirroring the `DBSPJoinBaseOperator` / `DBSPWindowOperator` assertions from `issue6655b`/`issue6655c` to assert here that no window/join was introduced, so a future regression that lets `ARRAY_AGG` slip through the allow-list is caught structurally rather than only by the numeric output.

**Review by mythical-fred — APPROVED** [review]
Nice, focused optimization with careful reasoning and thorough tests. The empty-group NULL/COUNT semantics, the `IS_NOT_DISTINCT_FROM` for grouping keys, the `emptyGroupValueKnown` allow-list, and the negative test for `ARRAY_AGG` all show the corner cases have been thought through. The build hygiene commit is cleanly separated.

A couple of minor items inline; approving.

**Review by gz — APPROVED** [review]
